### PR TITLE
Add support for contrib ports

### PR DESCRIPTION
**Pull request by ypujante**
Hi @sbc100 

This is a first draft of the work on contrib ports that we discussed in the https://github.com/emscripten-core/emscripten/pull/21148 PR

I used the PR that you authored a couple years ago https://github.com/emscripten-core/emscripten/pull/13002 as a basis for the work.

The point of this (draft) PR is to make sure that so far I have not broken anything and to give you a chance to chime in if you have any feedback. I would definitely love that.

The main idea (inspired by your work) is to allow contrib ports in the form of a python script under `tools/ports/contrib`.

* due to the fact that they are contrib ports, they have different requirements, like providing a project url / description and license. This extra information can then be used to automatically generate the documentation as you can see in this PR which is pretty cool
* a contrib port is enabled via the new setting: `USE_CONTRIB_PORT` which is a list since in theory you could have more than one
* I added a README in the contrib port folder with some details which amounts to: add a python script to this folder, regenerate the documentation... which I think would have very little burden on emscripten as a whole
* I did not change the main documentation to talk about this concept but I would if this PR would become a reality obviously

I am debating on whether the setting should be:

* `USE_CONTRIB_PORT` and then used this way: `-sUSE_CONTRIB_PORT=glfw3`
* `USE_PORT` and then used that way `-sUSE_PORT=contrib.glfw3` which opens up the door to extend the syntax to other built-in ports to be used this way as well, like `-sUSE_PORT=bzip2` (we would keep `-sUSE_BZIP2` syntax as well for legacy reasons and maybe deprecate it eventually)

Digging in this code, I feel like the whole port concept (not just contrib) could be revamped a bit:

* all ports are loaded whether any port is used or not every time emcc is invoked which is a waste: it requires reading/parsing all the python files in the folders, which is not insignificant. I believe we could be a lot more clever about this and load only the necessary ports. This would benefit everybody I believe. I can definitely work on that
* I would like to add a way to provide "options", which are equivalent to variants but using non built-in settings, maybe with a syntax like: `USE_PORT_OPTION=contrib.glfw3:DISABLE_WARNING`

Finally I have not looked at the url syntax yet because this might not be needed if contrib ports are allowed and easy to integrate.

I think the main/general idea in allowing external/contrib ports is that when somebody submit a new port it has minimal impact on the emscripten project from a review/change point of view (adding a file to folder/generating the documentation), and should have minimal/no impact on the runtime which I believe can be achieved if I work on loading ports only when necessary

### CODE REVIEW

**Comment by sbc100** [comment]
Regarding " it requires reading/parsing all the python files in the folders, which is not insignificant.".   Are you sure about that?  Can you measure the difference between running `./emcc hello_world.c` as it is today and without the ports tree being read?

File: tools/ports/contrib/contrib_example.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2024 The Emscripten Authors.  All rights reserved.
+# Emscripten is available under two separate licenses, the MIT license and the
+# University of Illinois/NCSA Open Source License.  Both these licenses can be
+# found in the LICENSE file.
+
+
+# The purpose of this contrib port is to have 2 contrib ports in order to have a test with
+# 2 ports since the syntax is different. Once there are 2 contrib ports available, this
+# example can be removed and the test changed to use the 2 available ports...
```
Comment (sbc100) [inline tools/ports/contrib/contrib_example.py:9]: How about using the libarchive example I provided in my PR?

File: tools/settings.py [hunk 2]
```diff
@@ -52,6 +52,7 @@
     'SDL2_MIXER_FORMATS',
     'SDL2_IMAGE_FORMATS',
     'USE_SQLITE3',
+    'USE_CONTRIB_PORT'
```
Comment (sbc100) [inline tools/settings.py:55]: Since settings are not (currently) additive, should this new setting be `-sUSE_CONTRIB_PORTS=` or `-sUSE_PORTS=`?

**Review by sbc100 — COMMENTED** [review]
Thanks for picking this up!   I'm glad to get that landed after all this time.

**Comment by sbc100** [comment]
If we are going to include your new glfw ports here I wonder if we could give it a unique name to distinguish it from the existing JS port?

**Comment by ypujante** [comment]
Thank you for your feedback.

> Regarding " it requires reading/parsing all the python files in the folders, which is not insignificant.". Are you sure about that? Can you measure the difference between running ./emcc hello_world.c as it is today and without the ports tree being read?

No I did not measure. But when you think about it, every time the build command is invoked it loads (right now) 25 python files/modules even if not a single port is used. I certainly don't believe it is scalable (long term) since the number of ports will most likely increase.

 > How about using the libarchive example I provided in my PR?

That was my first thought. The issue is that the library has evolved quite a bit since then and I certainly could not copy/paste your code. Given this is a draft PR/exploration phase I did not want to spend the extra time ATM to investigate this. That being said, if this gets finalized/approved I would not mind taking a deeper look at it to add a real other contrib as opposed to some random example.

> If we are going to include your new glfw ports here I wonder if we could give it a unique name to distinguish it from the existing JS port?

I am fine with that. I just don't know what it could be. `glfw3_cpp`? Definitely open to suggestions

> Thanks for picking this up! I'm glad to get that landed after all this time.

Happy to do it. Your PR definitely gave me a good starting point. The overall code is quite readable/understandable so it is actually quite a fun project. Especially if I can improve things in the end.

File: tools/settings.py (hunk 2 above)
Comment (ypujante) [reply tools/settings.py:55]: Yes I agree. I will use `-sUSE_PORTS`.

Comment (sbc100) [reply tools/settings.py:55]: Or maybe just `-sPORTS=`?

Comment (ypujante) [reply tools/settings.py:55]: Love it!

**Review by sbc100 — COMMENTED** [review]
I think this PR can probably not be split up.

1. Read ports on demand.   I'm still not convinced this is worth the complexity, its seems insignificant compared to the cost of either compiling or linking which is what emcc almost always does.   However, if we do do this it should be a separate change .

2. Add `-sPORTS` setting as alternative to `-sUSE_XXX` settings

3. Add `contrib` ports system.

4. Add PORT_OPTIONS.  I'm not sure I love this extra complexity.  Perhaps we can start by adding ports that don't require this, and then consider it later?

**Comment by ypujante** [comment]
I was in the process of explaining myself but you did not give me a chance to do so...

**Pull request closed without merging**